Added Amazon S3

MIRONTER opened the pull request.

MIRONTER committed:
Amazon S3

MIRONTER committed:
Update AmazonS3.cs

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace CViewer
+{
+    internal class AmazonS3
```
Добавь sealed модификатор. 
1. Это помогает компилятором Roslyn более оптимизированный код компилировать
2. Для архитектуры полезно, т.к. указываем, что больше у этого класса не будет наследников

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+
+        public List<string> GetFiles()
+        {
+            using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
```
Через alt+enter  (используя хоткеи решарпер, советую установить Resharper, если его нет) можно легко сделать явный тип. 

Явный тип предпочтительнее, так легче читать код. Как минимум в будущем стоит стараться делать явный тип (если он не супер-длинный со сложным написанием типов)

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+        {
+            using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
+            {
+                var list = client.ListObjects(
```
Из названия непонятно, что это за список. Хорошо бы сделать понятное имя.

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+                        BucketName = bucketName
+                    });
+
+                List<string> files = new List<string>();
```
Это filesNames, а не файлы. Или filesPaths (не знаю что конкретно возвращает амазон)

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+        {
+            using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
+            {
+                var list = client.ListObjects(
```
Это не Disposable объект? Если да, то надо обернуть в using.

ArtyomAfanasov replied:
@MIRONTER , ответь, пожалуйста, это всё-таки не Disposable? И далее, когда смотришь и решаешь комменты: либо код меняется, либо отвечаешь сам в обсуждение

ArtyomAfanasov replied:
Я посмотрел, не диспосбл. Но второе предложение предыдущего сообщения остается валидным :)

![image](https://user-images.githubusercontent.com/28110318/205974636-be59a2cd-d418-4f99-814a-f8effc7916fe.png)

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+
+                List<string> files = new List<string>();
+
+                foreach (var file in list.S3Objects)
```
`file` -- это не Disposable объект? Если да, то надо обернуть в using.

ArtyomAfanasov replied:
https://github.com/C-Viewer/CViewer/pull/10#discussion_r1041214222

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+    {
+        private readonly string bucketName = "cviewercvs";
+        private string accessKey;
+        private string secretKey;
```
Тоже можно `readonly`, как и `accessKey`.

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,92 @@
+                    Key = path
+                };
+
+                GetObjectResponse myResponse = client.GetObject(request);
```
**Важный коммент**:

Если путь некорректный, то, кажется, будет какое-то исключение. Надо обработать такой тип ошибки, чтобы не свалится, т.к., возможно `myResponse` даже может быть null или просто невалидным объектом.

ArtyomAfanasov commented:
Добавить try-catch на методы, которые работают с сетью

MIRONTER committed:
making edits

sealed класс
readonly ключи
объекты list и file не Disposable
Добавлено отлов ошибок, связанных с сетью

ArtyomAfanasovBS commented:
Ты не подключил nuget-пакет в наш проект. Я спуллил изменения, проект не билдится

![image](https://user-images.githubusercontent.com/79638564/205968306-baf9f7e8-0c3d-427b-be36-93e2a54a2d5c.png)

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -7,11 +7,11 @@
 namespace CViewer
 {
-    internal class AmazonS3
+    sealed internal class AmazonS3
```
Порядок модификаторов таков: `sealed internal`

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -7,11 +7,11 @@
         private readonly string bucketName = "cviewercvs";
-        private string accessKey;
-        private string secretKey;
+        private readonly string accessKey;
```
Давай private поля будем именовать, начиная с нижнего подчеркивания: _accessKey. И другие поля тоже

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -25,67 +25,97 @@ public AmazonS3(string access, string secret)
-                    {
-                        BucketName = bucketName
-                    });
+                using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
```
Переменные тут и везде всё ещё `var`. Давай будем указывать конкретные типы. Visual Studio или Resharper могут автоматически изменять var y на тип, используя горячие клавиши.

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -25,67 +25,97 @@ public AmazonS3(string access, string secret)
-                    });
+                using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
+                {
+                    var list = client.ListObjects(
```
Что за list? Комментарий не учтён https://github.com/C-Viewer/CViewer/pull/10#discussion_r1038981600

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -25,67 +25,97 @@ public AmazonS3(string access, string secret)
+            }
+            catch
+            {
+                return null;
```
Нужно завести класс (можно статический для скорости написания кода) для логирования и написать туда что произошло. И вызвать его тут и в других `catch`'ах данного класса

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,122 @@
+using Amazon.S3.Model;
+using System;
+using System.Collections.Generic;
+using System.IO;
```
Лишние юсинги отсюда нужно убирать

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,122 @@
+            accessKey = access;
+            secretKey = secret;
+            amazonS3Config.RegionEndpoint = RegionEndpoint.GetBySystemName("eu-north-1");
+            amazonS3Config.SignatureVersion = "4";
```
Почему, кстати, 4? Было бы хорошо сделать константу и назвать так, какой смысл несет эта цифра

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,122 @@
+        {
+            accessKey = access;
+            secretKey = secret;
+            amazonS3Config.RegionEndpoint = RegionEndpoint.GetBySystemName("eu-north-1");
```
Нужно вынести `"eu-north-1"` в хорошо именованную константу

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,122 @@
+            {
+                using (var client = AWSClientFactory.CreateAmazonS3Client(accessKey, secretKey, amazonS3Config))
+                {
+                    var list = client.ListObjects(
```
Нужно у try-catch сделать мЕньшую область действия. Только там, где можем упасть. Чуть ниже по коду в данном методе мы уже работаем без сети

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -0,0 +1,122 @@
+
+        public bool AddFile(FileStream stream, string path)
+        {
+            try
```
Тут и в других методах тоже уменьшить область try-catch

MIRONTER committed:
making edits #2

Extra usings are removed
internal sealed class
private fields are named starting with the underscore
constants are taken out
var replaced with an explicit type

MIRONTER committed:
Added Amazon endpoints

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    internal static class AmazonEndpoints
+    {
+        public static void MapAmazonEndpoints(this WebApplication app)
```
Ты не добавил использование этого метода в Program.cs MapAmazonEndpoints. В API сейчас не подключен амазон, получается. Значит ли это, что ты не тестировал работоспособность данной функциональности?

ArtyomAfanasov commented on Services/IAmazonS3Service.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface IAmazonS3Service
+    {
+        public List<string> GetFiles();
```
У тебя нет реализации данного интерфейса. Амазон-API точно не работает в данный момент

ArtyomAfanasov replied:
Надо сделать `AmazonS3Service : IAmazonS3Service`.

ArtyomAfanasov replied:
И важно! Проверить апишку через сваггер

ArtyomAfanasovBS committed:
Merge branch 'first-fix-for-api' into database

ArtyomAfanasov commented on DataAccess/AmazonS3.cs:
```diff
@@ -54,11 +53,11 @@ public bool AddFile(FileStream stream, string path)
-                        BucketName = bucketName,
+                        BucketName = _bucketName,
                         CannedACL = S3CannedACL.PublicRead,
                         Key = path,
```
**Важно**
Нет проверки на уникальность. Подозреваю, что мы не сможем добавить два файла с одинаковым `path`

ArtyomAfanasov replied:
Если это правда, то надо добавить проверку на уникальность на сервере. + генерить уникальное название

ArtyomAfanasov replied:
Прелагаю делать так:
Влад будет кидать в качестве параметра название файла. На сервер амазона нужно загружать такой ключ: `{имейл юзера}_{имя файла}_{номер повторяющегося файла с тем же именем}`.

ArtyomAfanasov replied:
Видимо, в CV_event надо добавить имя файла, чтобы Влад мог получать оригинальное имя. А GetFile будет по Id'шнику CV_event'а брать путь и качать нужный файл с Амазон

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,52 @@
+                    (IAmazonS3Service service) => GetFiles(service))
+                .Produces<List<string>>();
+
+            app.MapPost("/add_file",
```
Для каждого метода амазон-апи сделай проверку факта авторизации и проверку токена. Как это сделано например для метода `app.MapGet("/get_profile",`.
То есть навесить аттрибуты 
```
[EnableCors(Configuration.CorsPolicyName)]
                    [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
```
Добавит параметр контекста и сервис по безопасности. Также потом использовать сервис безопасности так:
```
if (!securityService.CheckAccess(applicantOrExpertToken))
            {
                return Results.Unauthorized();
            }
```

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,52 @@
+                .Produces<bool>();
+
+            app.MapGet("/get_file",
+                    (string path, IAmazonS3Service service) => GetFile(path, service))
```
добавить cv_event_id параметр

ArtyomAfanasov replied:
> Не нужно, из UI придет и так достаточно инфы

Не нужно, из UI придет и так достаточно инфы

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,52 @@
+                .Produces<Stream>();
+
+            app.MapGet("/delete_file",
+                    (string path, IAmazonS3Service service) => DeletetFile(path, service))
```
добавить cv_event_id параметр

ArtyomAfanasov replied:
> добавить cv_event_id параметр

Не нужно, из UI придет и так достаточно инфы

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,52 @@
+        public static void MapAmazonEndpoints(this WebApplication app)
+        {
+            app.MapGet("/get_all_files",
+                    (IAmazonS3Service service) => GetFiles(service))
```
это GetFileNames должно называться

MIRONTER committed:
making edits #3

MIRONTER committed:
Merge branch 'first-fix-for-api' into database

MIRONTER committed:
Merge branch 'first-fix-for-api' into database

MIRONTER committed:
some amazon bugs fix

MIRONTER committed:
Fixed all bugs

MIRONTER committed:
Deleted comments

MIRONTER committed:
Deleted the remaining comments...

ArtyomAfanasov commented on CViewer.csproj:
```diff
@@ -14,7 +16,9 @@
     <PackageReference Include="Microsoft.IdentityModel.Tokens" Version="6.14.1" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.2" />
     <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
+    <PackageReference Include="System.Configuration.ConfigurationManager" Version="7.0.0" />
```
Для чего этот пакет?

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    internal static class AmazonEndpoints
+    {
+
```
Лишняя строка

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+                .Produces<List<string>>();
+
+            app.MapPut("/add_file",
+                [EnableCors(Configuration.CorsPolicyName)]
```
Форматироваие отсупов шалит. Для простоты фикса: удали точку с запятой в данной строке и снова поставь. Вижуалка должна сама отформатировать строки красивее

ArtyomAfanasov replied:
В других местах тоже можно поправить это

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            app.MapGet("/get_file",
+                [EnableCors(Configuration.CorsPolicyName)]
+            [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+            ([Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => GetFile(path, context, securityService, service))
```
Лучше назвать fileKeyInAmazonStore, а не path. Будет сразу понятно при чтении кода что за параметр

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+
+        public static void MapAmazonEndpoints(this WebApplication app)
+        {
+            app.MapGet("/list_all_files",
+                    (IAmazonS3Service service) => GetFileNames(service))
+                .Produces<List<string>>();
```
Назови лучше list_all_cv_file_names-test:
1. test -- надо бы явно отделять методы, которые не нужны Владу в API
2. Тут ведь названия, а не сами файлы

Также соответственно надо поменять название исполняющего метода

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+                    (IAmazonS3Service service) => GetFileNames(service))
+                .Produces<List<string>>();
+
+            app.MapPut("/add_file",
```
Лучше назвать store_file

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            app.MapPut("/add_file",
+                [EnableCors(Configuration.CorsPolicyName)]
+            [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+            ([Required] IFormFile stream, [Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => AddFile(stream, path, context, securityService, service))
```
Слишком длинные строки, не помещающиеся на экран лучше разделять на несколько строк. Например, новую строку после "=>"

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            ([Required] IFormFile stream, [Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => AddFile(stream, path, context, securityService, service))
+                .Produces<bool>();
+
+            app.MapGet("/get_file",
```
Лучше назвать `get_serialized_file_stream`, мне кажется

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            app.MapDelete("/delete_file",
+                [EnableCors(Configuration.CorsPolicyName)]
+            [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+            ([Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => DeletetFile(path, context, securityService, service))
```
fileKeyInAmazonStore, а не path.

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            ([Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => GetFile(path, context, securityService, service))
+                .Produces<string>();
+
+            app.MapDelete("/delete_file",
```
Стоит переименовать в delete_file-Test

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+            ([Required] IFormFile stream, [Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => AddFile(stream, path, context, securityService, service))
+                .Produces<bool>();
+
```
Добавить метод `get_url_to_cv_file`, как мы обсудили в телеге

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+        private static IResult GetFileNames(IAmazonS3Service service)
+        {
+            Task<List<string>> status = service.GetFileNamesAsync();
+            status.Wait();
```
Всё преимуществно асинхронного метода амазона уничтожается данной строкой. Код выходим синхронным и ожидает в том же потоке.

Для асинхронности привожу пример кода для данного метода. По аналогии надо сделать все остальные методы также асинхронными

```
        private static async Task<IResult> GetFileNames(IAmazonS3Service service)
        {
            List<string> allCVFileNames = await service.GetFileNamesAsync();
            return Results.Ok(allCVFileNames);
        }
```

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+        private static IResult AddFile(IFormFile stream, string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service)
+        {
+            string token = TokenHelper.GetToken(context);
+            if (!securityService.CheckAccess(token)) { return Results.Unauthorized(); }
```
Здесь и в других местах для единообразия кода лучше делать переносы скобок для всех if'ов:
```
 if (!securityService.CheckAccess(token))
            {
                return Results.Unauthorized();
            }
```

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+                [EnableCors(Configuration.CorsPolicyName)]
+            [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+            ([Required] IFormFile stream, [Required] string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service) => AddFile(stream, path, context, securityService, service))
+                .Produces<bool>();
```
Стоит убрать этот `.Produces<bool>();`, т.к. метод будет возвращать либо `Besults.BadRequest("File upload failed");`. либо `Results.Ok()`

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            if (!securityService.CheckAccess(token)) { return Results.Unauthorized(); }
+            Task<string> status = service.GetFileAsync(path);
+            status.Wait();
+            if (status == null) return Results.NotFound("File not found");
```
Проверять на null нужно `status.Result`, а не статус. Но т.к. ты изменишь код на асинхронный, то проверят на null нужно будет получающийся результат

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            Task<bool> status = service.AddFileAsync(stream, path);
+            status.Wait();
+            if (!status.Result) return Results.BadRequest("File upload failed");
+            return Results.Ok(true);
```
Возвращать просто `Results.Ok()`

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -0,0 +1,75 @@
+            Task<bool> status = service.DeleteFileAsync(path);
+            status.Wait();
+            if (!status.Result) return Results.NotFound("File not found");
+            return Results.Ok(true);
```
Просто `Results.Ok()`

MIRONTER committed:
making edits #4

ArtyomAfanasov commented on Endpoints/AmazonEndpoints.cs:
```diff
@@ -9,67 +9,73 @@ namespace CViewer.Endpoints
         }
 
-        private static IResult DeletetFile(string path, HttpContext context, ISecurityService securityService, IAmazonS3Service service)
+        private static async Task<IResult> DeletetFileAsync(string fileKeyInAmazonBucket, HttpContext context, ISecurityService securityService, IAmazonS3Service service)
```
Буква t лишняя

ArtyomAfanasov commented on Services/AmazonS3Service.cs:
```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        public async Task<bool> AddFileAsync(IFormFile stream, string path)
```
Лучше в параметре int cv_event_id, а не string path, чтобы UI не путался и точно подавал id-шник ивента. А тут конвертнуть его в строку

MIRONTER committed:
Deleted t in Deletet

MIRONTER committed:
Update CViewer.csproj
